lib/mix/tasks/firmware.burn.ex

```diff
@@ -160,4 +199,24 @@ defmodule Mix.Tasks.Firmware.Burn do
     gb = bytes / 1024 / 1024 / 1024
     Float.round(gb, 2)
   end
+
+  defp is_wsl? do
+    #using system.cmd("cat", ...) here is simpler
+    #https://stackoverflow.com/questions/29874941/elixir-file-read-returns-empty-data-when-accessing-proc-cpuinfo/29875499
+    System.cmd("cat", ["/proc/sys/kernel/osrelease"]) 
+      |> elem(0)
+      |> (&Regex.match?(~r/Microsoft/, &1)).()
+  end
+
+  defp get_wsl_paths(file) do
+    {win_path, 0} = System.cmd("pwd.exe", [])  
```

Comment:
This line fails for me with the following error. I'm curious now, because within WSL I have pwd but not pwd.exe. 

Just to clarify, is the expectation is that `mix firmware.burn` is supposed to be run from within wsl, and it shells out to PowerShell to execute Windows tasks? That's my impression but doesn't match with the pwd.exe change here.

```
** (ErlangError) Erlang error: :enoent
    (elixir) lib/system.ex:622: System.cmd("pwd.exe", [], [])
    lib/mix/tasks/firmware.burn.ex:212: Mix.Tasks.Firmware.Burn.get_wsl_paths/1
    lib/mix/tasks/firmware.burn.ex:65: Mix.Tasks.Firmware.Burn.run/1
    (mix) lib/mix/task.ex:314: Mix.Task.run_task/3
    (mix) lib/mix/cli.ex:80: Mix.CLI.run_task/2
```

Reply:
I'm not sure why you are unable to run that command. You should be able to run any windows command by adding a '.exe' to the end: https://docs.microsoft.com/en-us/windows/wsl/interop#run-windows-tools-from-wsl. What happens when you launch WSL and just try to run 'pwd.exe' or something like that? I just noticed this, if I run 'pwd.exe', it works fine, but 'sudo pwd.exe' gives me a 'command not found' error. Maybe that's why it is not working.

Yes, it is supposed to run within WSL and launch powershell and windows tasks... It is very spaghetti... I'm going to attempt to explain why I am doing certain things that don't seem to make much sense.

The reason I implemented it this way is that I could not get fwup to work within WSL. Windows does not seem to allow WSL to access the SD card devices. However, it is possible to install fwup.exe via chocolately, and running fwup.exe from WSL seemed to work fine.

BUT - fwup.exe still will not work unless the program is elevated (running as administrator). I use powershell to launch an elevated process to do this. You may think, 'josh, why didn't you just relaunch the entire script elevated?' Well, I tried to do just that, but in my version of windows, it is impossible to launch an unelevated and an elevated WSL instance at the same time. They added support for this in a newer build, but I am unfortunately unable to access that build: https://blogs.msdn.microsoft.com/commandline/2017/12/04/background-task-support-in-wsl/. 

So, where we are now: we can burn the firmware by using powershell to launch an elevated instance of fwup.exe that is able to access the devices. The next problem I encountered was introduced by you six days ago: it only worked if windows is able to access the project directory! Because fwup.exe is technically a windows program, it can't access of the project files if they are in the WSL filesystem.

They way I overcame that is by temporarily copying the files fwup.exe would need into directories windows is able to access. The way I decided what directory to copy to is by running 'pwd.exe' within WSL. If you are working within the windows file system, pwd.exe outputs the projects directory, which is a good place for temp files. If you are working within the WSL filesystem, pwd.exe will output the path to your windows home directory '/c/Users/username'. That is the directory I used to store the temporary files.

For getting the list of available devices, I do pretty much the same thing. I use powershell to launch 'fwup.exe --devices' and pipe the output of that into a file in a directory that is accessible by both windows and WSL. Then I remove the file afterwords.

Ok... I think I explained all the crazy stuff I had to do to get this to work. If you have any more questions, please let me know. If you think of a much better way to accomplish this, PLEASE let me know. And thanks for helping me improve my code, I appreciate it.

Reply:
Turns out 'pwd.exe' was only working on my machine because it was installed with some other program (probably git for windows) and does not work on normal windows machines. I changed it to be 'cmd.exe /c cd', which should work on every windows machine. Please let me know if it is still not working for some reason.